codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.Arrays;
+
+/**
+ * A {@link StreamByteAssigner} that implements the HTTP/2 priority tree algorithm for assigning
+ * bytes for all streams in the connection.
+ */
+public final class PriorityStreamByteAssigner implements StreamByteAssigner {
+    private final Http2Connection connection;
+    private final Http2Connection.PropertyKey stateKey;
+    private final StreamVisitor visitor = new StreamVisitor();
+    private int totalAssigned;
+
+    public PriorityStreamByteAssigner(Http2Connection connection) {
+        this.connection = checkNotNull(connection, "connection");
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        connection.connectionStream().setProperty(stateKey,
+                new PriorityState(connection.connectionStream()));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new PriorityState(stream));
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+
+            @Override
+            public void onPriorityTreeParentChanged(Http2Stream stream, Http2Stream oldParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    int delta = state(stream).unassignedStreamableBytesForTree();
+                    if (delta != 0) {
+                        state(parent).unassignedStreamableBytesForTreeChanged(delta);
+                    }
+                }
+            }
+
+            @Override
+            public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    int delta = state(stream).unassignedStreamableBytesForTree();
+                    if (delta != 0) {
+                        state(parent).unassignedStreamableBytesForTreeChanged(-delta);
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void update(StreamState streamState) {
+        state(streamState.stream()).update(streamState);
+    }
+
+    @Override
+    public boolean assign(int maxBytes) {
+        if (maxBytes > 0) {
+            assignBytesForTree(connection.connectionStream(), maxBytes);
+        }
+        return state(connection.connectionStream()).unassignedStreamableBytesForTree() -
+                totalAssigned > 0;
+    }
+
+    @Override
+    public void processAssignedBytes(Visitor visitor) {
```

Comment:
`visitor` is used a few times here. Consider renaming the `this.visitor` to `streamVisitor` and the `visitor` arg to `assignmentVisitor`?